### ISSUE TITLE
OpenID Connect (OIDC) authentication support

### ISSUE THREAD

**Issue by bzub** [siderolabs/talos#3306]
### Description

We use oidc to map LDAP/Active directory users/groups to RBAC resources for kubernetes auth.
https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens

Would be nice if we could do the same for Talos API, even if it's not as granular as RBAC (admin only).

**Comment by Ulexus** [comment]
This should probably plug into dex, rather than building it out ourselves.

**Comment by devopstales** [comment]
Hi @Ulexus. I think you misunderstand this ticket. The kubernetes api-server has an ability to authenticate users not just wit certificates but wit JWT tokens. This token is generated by an OpenID authentication provider (the authentication provider  can be Dax, keycloak .. etc). To do the authentication the kubernetes api-server needs a configuration to know where the authentication provider and how to use it. This ticket is about that configuration, and it is valid because at the moment there is no option in altos to add this kind of configuration, or just not documented.

**Comment by Ulexus** [comment]
@devopstales Is there something lacking right now?  You can pass all of the OIDC configuration options to kube-apiserver via `extraArgs`:  https://www.talos.dev/v1.0/reference/configuration/#apiserverconfig

**Comment by wrmedford** [comment]
I've started implementing this in some other projects (ArgoCD example [here](https://github.com/argoproj/argo-cd/pull/20928) which also uses Dex). I'm also work at [Pomerium](https://github.com/pomerium/pomerium), so I'm looking at if it would make sense to pair us as an extension to Talos similar to how there's one for Tailscale, that way you could offload OIDC for both Talos and Kubernetes to a proxy that would either run on host or externally. Would help with cold start security posture as well, as well as keeping secrets out of disk images, as you can bake in JWKS configs without any of it being secret. 

Would be happy to pick up this issue if I could just get a hand getting oriented to the project!

**Comment by jenrik** [comment]
This is still a desired featured and should not be closed

**Comment by smira** [comment]
[Omni](https://siderolabs.com/omni/) provides this feature on top of Talos.

Talos API itself should stay simple and provide access even if the OIDC provider is down/misconfigured.

**Comment by WolfeCub** [comment]
> Talos API itself should stay simple and provide access even if the OIDC provider is down/misconfigured.

`talosconfig` files should work regardless of the OIDC provider state but it would be nice to have an alternative authentication option for those of us not using Omni.

**Comment by smira** [comment]
> > Talos API itself should stay simple and provide access even if the OIDC provider is down/misconfigured.
> 
> `talosconfig` files should work regardless of the OIDC provider state but it would be nice to have an alternative authentication option for those of us not using Omni.

If you don't like Omni, you can build something for it, but we certainly don't want to have parallel implementations for the same feature.

**Comment by WolfeCub** [comment]
Definitely understand not wanting to maintain two implementations of the same thing. Ideally it would be great if the implementation for OIDC "just" moved to talos and then Omni made use of that feature.

However I get wanting to keep that exclusive to Omni.

**Comment by smira** [comment]
> Definitely understand not wanting to maintain two implementations of the same thing. Ideally it would be great if the implementation for OIDC "just" moved to talos and then Omni made use of that feature.
> 
> However I get wanting to keep that exclusive to Omni.

I think there is misunderstanding here. OIDC is authentication, there is also authorization piece. Also you need "landing page" for OIDC. Omni does all of that, it doesn't make much sense to be in Talos, it should be external to Talos system which trades OIDC tokens for proper proxied access to Talos without leaking `talosconfig`. 

Omni does much more on top of that, e.g. offering tunneling over SideroLink. There is no place for it in Talos itself, that's why it's a separate component outside of the cluster, in fact shared across all your clusters.

**Comment by WolfeCub** [comment]
> I think there is misunderstanding here. OIDC is authentication, there is also authorization piece. Also you need "landing page" for OIDC. Omni does all of that, it doesn't make much sense to be in Talos, it should be external to Talos system which trades OIDC tokens for proper proxied access to Talos without leaking talosconfig.

Right. For context I'm currently using https://github.com/int128/kubelogin to gain access to my cluster. When logging in I'm directed to my provider in the browser for login and then given a token that I can then use for access to the cluster. I'm really just looking for something similar for `talosctl` a way to gain access to the cluster via an external provider rather than having to manage talosconfigs on multiple machines.

**Comment by rochecompaan** [comment]
I came here with the same question. We're also using kubelogin, and securing talos api access with SSO was the final step in securing our clusters. We were considering writing something to issue short-lived certificates, but I just came across this effort, which looks promising: https://github.com/qjoly/talosctl-oidc.

**Comment by steverfrancis** [comment]
closing this as not planned for architectural reasons.

A proper OIDC integration is auth + authorization + a browser-redirect landing page + token exchange. Putting that machinery inside the Talos API server would compromise the "minimal, always-available host OS" property that's most of Talos's value. The correct architectural layer for OIDC is external — a component that sits in front of Talos, accepts OIDC tokens, and proxies authenticated access to the Talos API without requiring talosconfig to be distributed.
For the underlying need most of you described — SSO-fronted talosctl access so admins aren't passing talosconfig files between machines — there are two supported paths:

- Omni ([siderolabs.com/omni](https://siderolabs.com/omni)) is the productized answer. Omni handles OIDC auth, RBAC, audit logs, and SideroLink tunneling to clusters; omnictl and the web UI replace talosconfig distribution. Available as SaaS or self-hosted.
- An external OIDC proxy for users who don't want Omni. @qjoly's [talosctl-oidc](https://github.com/qjoly/talosctl-oidc) — community project demonstrating the model.

We'd reopen this issue if an enterprise customer attaches OIDC-in-core to a paid engagement, or if a mature community-built external proxy emerges and the community wants Sidero to bless or steward it.
 If discussion of external-proxy approaches would be useful, we're happy to convert this to a GitHub Discussion rather than leave a closed issue as the venue.

**Comment by christiaan-bronkhorst** [comment]
@steverfrancis how would a mature community-built solution look like? If something like this https://github.com/qjoly/talosctl-oidc already exist, why not just fork it? I'm willing to help out on this https://github.com/qjoly/talosctl-oidc project to get it to the maturity you require it to be. Not sure if there are other alternate projects that already exist that would be preferred over the one I listed.

**Comment by michaelbeaumont** [comment]
> Talos API itself should stay simple and provide access even if the OIDC provider is down/misconfigured.
> it doesn't make much sense to be in Talos
> Putting that machinery inside the Talos API server would compromise the "minimal, always-available host OS" property that's most of Talos's value.
> The correct architectural layer for OIDC is external

Hmm is this also the case for Kubernetes and its JWT token auth support? As mentioned in other comments, things like "browser-redirect landing page, token exchange" happen _in the client_ with k8s + OIDC. Hmm, it's actually almost all pretty "external" from the k8s API. What makes the Talos control plane so different from k8s that it "shouldn't" support the same JWT token auth mechanism that enables things like `kubelogin`, which is exactly what people are asking for?

**Comment by smira** [comment]
> Hmm is this also the case for Kubernetes and its JWT token auth support? As mentioned in other comments, things like "browser-redirect landing page, token exchange" happen _in the client_ with k8s + OIDC. Hmm, it's actually almost all pretty "external" from the k8s API. What makes the Talos control plane so different from k8s that it "shouldn't" support the same JWT token auth mechanism that enables things like `kubelogin`, which is exactly what people are asking for?

The thing is that is your OIDC provider changes settings, you can reconfigure Kubernetes by going via the OS and updating Kubernetes config. Talos API is the _only_ way to access Talos Linux, so if you lose access, you can't recover it.

This is why Talos API access is easy and straightforward. If you need external access via OIDC/..., you can build a Talos API proxy with that.

In fact, we built one - it's one of the [Omni](https://www.siderolabs.com/omni) features, among many others. Core Talos API will stay minimal, OIDC auth can be handled outside of Talos API.

**Comment by michaelbeaumont** [comment]
> The thing is that is your OIDC provider changes settings, you can reconfigure Kubernetes by going via the OS and updating Kubernetes config. Talos API is the only way to access Talos Linux, so if you lose access, you can't recover it.

I might be missing something but couldn't the API just allow both mTLS and JWT [like k8s does](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authentication-methods)?

> the Omni features

Omni is then a thing you need to run somewhere. Omni is great but it is super overkill for the one thing compared to just... authenticating JWT tokens. And again, not sure about the latest Omni features/integrations but users running on Terraform/Cluster API are kinda stuck.

**Comment by smira** [comment]
> Omni is then a thing you need to run somewhere. Omni is great but it is super overkill for the one thing compared to just... authenticating JWT tokens. And again, not sure about the latest Omni features/integrations but users running on Terraform/Cluster API are kinda stuck.

Omni brings more than JWT. Terraform provider for Omni is coming soon, and Cluster API is really bad for Talos in-place update nature.

So we don't see any point in this in core Talos.
